### ISSUE TITLE
load_weights does not work when weights are saved in h5 format

### ISSUE THREAD

**Issue by bj1123** [tensorflow/tensorflow#35930]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.1
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**
model can't restore weights from h5 file

**Describe the expected behavior**

**Code to reproduce the issue**
```
class Model(keras.Model):
    def __init__(self, inp1, inp2):
        super(Model, self).__init__()
        self.x1 = self.add_weight('w1',[inp1])
        self.x2 = self.add_weight('w2',[inp2])
    def call(self,x):
        return x
# load_weights method works when save format is 'tf
x = Model(100,200)
x.save_weights('temp.tmp',save_format='tf')
old = x.weights[0][0].numpy()
print(old)
x = Model(100,200)
x.load_weights('temp.tmp')
new = x.weights[0][0].numpy()
print(new)
print(old==new)

# load_weights method does not work when save format is 'h5'
x = Model(100,200)
x.save_weights('temp.h5',save_format='h5')
old = x.weights[0][0].numpy()
print(old)
x = Model(100,200)
x.load_weights('temp.h5')
new = x.weights[0][0].numpy()
print(new)
print(old==new)
```
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by oanush** [comment]
@bj1123 ,
When tried running the given code in both [2.1](https://colab.sandbox.google.com/drive/1rqcqkeDpD0TG3brFJtyq1x8WEfET0vcU#scrollTo=RbZaVocMg-u_) and [tf-nightly](https://colab.sandbox.google.com/drive/19sCVnZyjsRYzxeCs5M5IjMGli_nP-PnK#scrollTo=0FawR_S5fFm8), I didn't face any error.Thanks!

**Comment by bj1123** [comment]
> @bj1123 ,
> When tried running the given code in both [2.1](https://colab.sandbox.google.com/drive/1rqcqkeDpD0TG3brFJtyq1x8WEfET0vcU#scrollTo=RbZaVocMg-u_) and [tf-nightly](https://colab.sandbox.google.com/drive/19sCVnZyjsRYzxeCs5M5IjMGli_nP-PnK#scrollTo=0FawR_S5fFm8), I didn't face any error.Thanks!

It does'nt raise error. The problem is that the saved weights are not restored.

**Comment by oanush** [comment]
@bj1123 ,
I just modified few lines of code and I am able to load the model weights successfully,kindly refer the [gist of colab](https://colab.sandbox.google.com/gist/oanush/8b8802b2743fbaf47149c5941da8e213/35930_tf-nightly.ipynb). Please refer these links of [github](https://github.com/keras-team/keras/issues/12094#issuecomment-481397688) and [stackoverflow](https://stackoverflow.com/questions/41859997/keras-model-load-weights-for-neural-net) for more information.

**Comment by bj1123** [comment]
@oanush 
I think your modification can not assure the weights are successfully loaded. If you comment out the code for initializing the model, there is no difference in the model weights. Printing old and new variable happen to be the same whether or not the weights are successfully loaded. I slightly changed the codes for better understanding, and it still shows the same problem.
```
class Model(keras.Model):
    def __init__(self, inp1, inp2):
        super(Model, self).__init__()
        self.x1 = self.add_weight('w1',[inp1])
        self.x2 = self.add_weight('w2',[inp2])
    def call(self,x):
        return x
# load_weights method works when save format is 'tf
x = Model(100,200)
x.save_weights('temp.tmp',save_format='tf')
old = x.weights[0][0].numpy()
print(old)
x.load_weights('temp.tmp')
new = x.weights[0][0].numpy()
print(new)
print(old==new)

# load_weights method does not work when save format is 'h5'
x = Model(100,200)
x.save_weights('temp.h5',save_format='h5')
old = x.weights[0][0].numpy()
x.weights[0].assign(tf.zeros_like(x.weights[0]))
print(old)
x.load_weights('temp.h5')
new = x.weights[0][0].numpy()
print(new)
print(old==new)
```

**Comment by mbelloc** [comment]
Any update regarding this issue ? 

I am having the same problem trying to load inceptionV3 weights trained with TF1.13  using TF2.1.

**Comment by k1moradi** [comment]
This issue is so annoying. I am using both Anaconda's TF 2.1 GPU and MKL on both Linux and Windows. After training loss is for example 0.2 loading the weights error increases significantly as if the model is not trained.

**Comment by cs-mac** [comment]
I also have this issue, anyone know a solution yet?

**Comment by bj1123** [comment]
@cs-mac 
'''

def load_weights(model,save_path):
    hf = h5py.File(save_path, 'r')
    for i in model.trainable_weights:
        res = hf.get(i.name)
        res = tf.convert_to_tensor(np.array(res))
        if res.shape == i.shape:
            i.assign(res)
'''

**Comment by cs-mac** [comment]
@bj1123 
Thank you for your reply! 
I managed to fix the mistake, embarrassingly enough it was a user error on my part.
I remade the whole trained model to access some layer I needed. Then I used this newly created model, which of course was not trained yet, instead of the trained model. Therefore, all attention weights seemed random.
